### PR TITLE
adds monitoring to complaint detail timeline

### PR DESCRIPTION
**Pull request by jduss4**
## Description of change

Displays linked monitoring reviews in an individual complaint's timeline alongside TTA activity and complaint updates.  I didn't see a design specifically for this so I just picked some fields that seemed like they would be nice to display. If we would prefer to take something like "status" and change it into verbage the way that the complaint updates look (in progress / etc) we could do that.

## Acceptance Criteria

 - add event type for monitoring review
 - incorporate into complaint detail view timeline
 - displays the correct information for the monitoring activity (currently chose status and type....not sure if that's right!)

## How to test

Fire up the complaint tracker, head to a complaint's page, add a RAN ID, and it should appear in the timeline.

image

## Issue(s)

* closes https://github.com/OHS-Hosting-Infrastructure/complaint-tracker/issues/204

## Checklist

To be completed by the submitter:

- [ ] Meets issue criteria
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

## To the Reviewer

This project is using [Conventional Comments](https://conventionalcomments.org/) to give structure
and context to PR comments. Please use these labels in your comments.

* **praise:**
* **nitpick:**
* **suggestion:**
* **issue:**
* **question:**
* **thought:**
* **chore:**

### CODE REVIEW

**Comment by jduss4** [comment]
This PR is in draft because @lauranash18f is going to look it over for design, but the bulk of the code seems unlikely to change (at least, I'm hoping) if people are looking for something to review

File: app/models/timeline.rb [hunk 1]
```diff
@@ -9,11 +9,12 @@ class Timeline
     initialContactDate: "Initial contact from complaint"
   }.with_indifferent_access.freeze
 
-  def initialize(complaint_attributes, tta_reports)
+  def initialize(complaint_attributes, tta_reports = [], monitoring_reviews = [])
```
Comment (Sgtpluck) [inline app/models/timeline.rb:12]: [suggestion] i don't think we need the default values; they are both being fetched with a `.where`, which returns an empty array if nothing is found.

File: app/models/timeline.rb [hunk 2]
```diff
@@ -25,7 +26,7 @@ def events
   private
 
   def all_events
-    @complaint_events + @tta_events
+    @complaint_events + @tta_events + @monitoring_events
```
Comment (Sgtpluck) [inline app/models/timeline.rb:29]: [praise] yay

File: app/models/timeline/monitoring_event.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    @date = issue_monitoring_review.start_date
+    @review = issue_monitoring_review.monitoring_review
+  end
+
+  def api_call_succeeded?
+    review.valid?
+  end
+
+  def review_link
+    # using content_tag for the link will properly escape any dangerous characters that sneak into name or ttahub_url
```
Comment (Sgtpluck) [inline app/models/timeline/monitoring_event.rb:18]: [nitpick] s/ttahub/itams

File: app/views/complaints/_monitoring_timeline_event.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<h3 class="usa-process-list__heading ct-timeline__two-part-container">
```
Comment (Sgtpluck) [inline app/views/complaints/_monitoring_timeline_event.html.erb:1]: [thought] we could *probably* better align the monitoring/ttahub events and reduce duplication, but *also* probably not worth it

File: spec/models/timeline/monitoring_review_spec.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
+require "rails_helper"
+
+RSpec.describe Timeline::MonitoringEvent do
+  let(:date) { 1.day.ago.strftime("%F") }
+  let(:review_id) { "Test-Review-ID" }
+  let(:issue_id) { "Test-Issue-ID" }
+  # just setting the start_date manually because we aren't triggering the before_validation callback
+  let(:event_param) { IssueMonitoringReview.new review_id: review_id, issue_id: issue_id, start_date: date }
+  let(:monitoring_review) { event_param.monitoring_review }
```
Comment (Sgtpluck) [inline spec/models/timeline/monitoring_review_spec.rb:9]: [suggestion] this doesn't look used, can we axe it?

File: spec/models/timeline/monitoring_review_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    end
+
+    context "authorization error loading Monitoring API" do
+      xit "returns just the review ID with no link"
+    end
+  end
+
+  describe "#review_type" do
+    it "returns the review type" do
+      expect(subject.review_type).to be_a(String)
```
Comment (Sgtpluck) [inline spec/models/timeline/monitoring_review_spec.rb:32]: [suggestion] can we have it match on the options? 
`expect(subject.review_type.include?(FfakerWrapper::REVIEW_TYPE)).to be true` (or something like that?)

File: spec/models/timeline/monitoring_review_spec.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    end
+  end
+
+  describe "#status" do
+    it "returns the review status" do
+      expect(subject.status).to be_a(String)
```
Comment (Sgtpluck) [inline spec/models/timeline/monitoring_review_spec.rb:38]: same as above comment

File: spec/models/monitoring_review_spec.rb [hunk 8]
```diff
@@ -8,12 +8,30 @@
...
+
   describe "#start_date" do
     it "returns a start date" do
       expect(subject.start_date).to be_a(String)
     end
   end
 
+  describe "#status" do
+    it "returns a status" do
+      expect(subject.status).to be_a(String)
```
Comment (Sgtpluck) [inline spec/models/monitoring_review_spec.rb:31]: [suggestion] can we have it match on the options?
`expect(subject.review_type.include?(FfakerWrapper::REVIEW_STATUS)).to be true` (or something like that?)

File: spec/models/monitoring_review_spec.rb [hunk 9]
```diff
@@ -8,12 +8,30 @@
     described_class.new id, access_token
   end
 
+  describe "#itams_url" do
+    it "returns a URL" do
+      expect(subject.start_date).to be_a(String)
```
Comment (Sgtpluck) [inline spec/models/monitoring_review_spec.rb:13]: [issue] this isn't testing the URL!

File: spec/models/monitoring_review_spec.rb [hunk 10]
```diff
@@ -8,12 +8,30 @@
...
+    end
+  end
+
+  describe "#review_type" do
+    it "returns the URL to view the report in IT-AMS" do
+      expect(subject.itams_url).to eq "https://example.com/TODO"
```
Comment (Sgtpluck) [inline spec/models/monitoring_review_spec.rb:19]: [issue] this isn't testing the review_type!

**Review by Sgtpluck — CHANGES_REQUESTED** [review]

File: app/models/timeline.rb (hunk 1 above)
Comment (jduss4) [reply app/models/timeline.rb:12]: That makes sense. I added them mostly because of the test suites trying to start up a timeline with only tta, but that's kind of an edge case so I might as well drop this

File: app/views/complaints/_monitoring_timeline_event.html.erb (hunk 4 above)
Comment (jduss4) [reply app/views/complaints/_monitoring_timeline_event.html.erb:1]: I just got a couple design pointers from Laura which are going to make them diverge just slightly more, plus they're pulling different fields.....basically I decided it wasn't worth it for the small amount of code in these files

File: spec/models/timeline/monitoring_review_spec.rb (hunk 5 above)
Comment (jduss4) [reply spec/models/timeline/monitoring_review_spec.rb:9]: Oooh I'll go check that out.  Maybe we can ditch the line I grabbed from the tta_review_spec, too

File: spec/models/monitoring_review_spec.rb (hunk 10 above)
Comment (jduss4) [reply spec/models/monitoring_review_spec.rb:19]: wow I really did a good job on this part

**Comment by jduss4** [comment]
@lauranash18f requests that outcome be displayed whenever the status is complete, so this is what we end up with

image

CONCERN 😟

File: spec/models/timeline/monitoring_review_spec.rb (hunk 5 above)
Comment (jduss4) [reply spec/models/timeline/monitoring_review_spec.rb:9]: TTA is using it but monitoring is not.  Nice catch!

File: app/models/timeline/monitoring_event.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def api_call_succeeded?
+    review.valid?
+  end
+
+  def review_link
+    # using content_tag for the link will properly escape any dangerous characters that sneak into name or monitoring url
+    content_tag(:a, name, class: "usa-link", href: review.itams_url, target: "_blank")
+  rescue Api::Error
```
Comment (rahearn) [inline app/models/timeline/monitoring_event.rb:21]: [question] where does this get raised?

**Review by rahearn — CHANGES_REQUESTED** [review]
I think the error raise is probably being punted to the error task, but I think an inline-comment would help make sure that implementation detail of raising the error that matches what the timeline event is expecting isn't forgotten

File: app/views/complaints/_monitoring_timeline_event.html.erb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  <span>Monitoring Activity: <%= event.review_link %></span>
+  <span class="ct-timeline__date"><%= event.formatted_date %></span>
+</h3>
+<% if event.api_call_succeeded? %>
+  <div>
+    <ul class="usa-list usa-list--unstyled">
+      <li><strong>Status:</strong> <%= event.status %></li>
+      <% if event.status == "Complete" %>
```
[the remark this answers is not in the document]
Comment (jduss4) [reply app/views/complaints/_monitoring_timeline_event.html.erb:9]: I like that idea!

File: app/models/timeline/monitoring_event.rb (hunk 11 above)
Comment (jduss4) [reply app/models/timeline/monitoring_event.rb:21]: Good catch, I had copied that over from the TTA activity report without thinking about how I pulled out the `get_api_data_field` error for monitoring. If you think just commenting it for now is the way to go, I can do that and I'll add a note on the error issue.

Comment (rahearn) [reply app/models/timeline/monitoring_event.rb:21]: Yeah, I think it's fine as long as the person implementing the error task is super aware of this being here

**Review by rahearn — APPROVED** [review]
🎉
